### DISCUSSION TITLE
Getting a owned IriRef<MownStr>> from a SimpleTerm

### DISCUSSION THREAD

**Discussion by GordianDziwis** [pchampin/sophia_rs#167] [Q&A]
I am currently converting my codebase to sophia 0.8.0. And I have some ugliness I want to get rid of. In the `Some(dt_inner)` match arm, I need an owned `IriRef<MownStr>>` from a `SimpleTerm`, which gets complicated because `iri()` takes a reference to self of `simple_term` which goes out of scope after the closure.

My workaround is:

```rust
                let dt = match datatype.as_ref() {
                    Some(dt_inner) => dt_inner.try_as_term(diagram).map(
                        |simple_term: SimpleTerm| {
                            IriRef::new_unchecked(
                                simple_term.iri().unwrap().to_string().into(),
                            )
                        },
                    )?,
                    None => xsd::string.iri().unwrap(),
                };
                SimpleTerm::LiteralDatatype(lexical_form.as_str().into(), dt)
```

Is there a nicer way to do this?

**Comment by pchampin** [comment]
You could remove some ugliness by using [`IriRef::map_unchecked`](https://docs.rs/sophia_iri/0.8.0/sophia_iri/struct.IriRef.html#method.map_unchecked).
```rust
        let dt = match datatype.as_ref() {
            Some(dt_inner) => dt_inner
                .try_as_term(diagram)?
                .iri()
                .unwrap()
                .map_unchecked(|txt| txt.to_string().into()),
            None => xsd::string.iri().unwrap(),
        };
        SimpleTerm::LiteralDatatype(lexical_form.as_str().into(), dt)
```

Assuming you have control over the type of `dt_inner`, you might also add a dedicated method to it (which would do exactly what is above), allowing you to write:
```rust
        let dt = match datatype.as_ref() {
            Some(dt_inner) => dt_inner.try_as_iri(diagram)?,
            None => xsd::string.iri().unwrap(),
        };
        SimpleTerm::LiteralDatatype(lexical_form.as_str().into(), dt)
```
where the return type of `try_as_iri(diagram)` is `IriRef<MownStr<'static>>`.

NB: if you don't have control over the type of `dt_inner`, you could still add the `try_as_iri` method via an [extension trait](https://rust-lang.github.io/rfcs/0445-extension-trait-conventions.html#what-is-an-extension-trait).
